rosidl_generator_cpp/include/rosidl_generator_cpp/MessageTypeSupport.h

```diff
@@ -17,13 +17,22 @@
 #define __rosidl_generator_cpp__MessageTypeSupport__h__
 
 #include <rosidl_generator_c/message_type_support.h>
+#include <rosidl_generator_c/visibility_control.h>
 
 namespace rosidl_generator_cpp
 {
 
 template<typename T>
+ROSIDL_PUBLIC
 const rosidl_message_type_support_t * get_type_support_handle();
 
 }  // namespace rosidl_generator_cpp
 
+// Tail include the implementation for this function.
+// This include is provided by the implementation and therefore only
+// one version of this header will get included from the path.
+// This header maps the get_type_support_handle function to an implementation
+// specific version of that function.
+#include <rosidl_generator_cpp/get_type_support_handle_impl.hpp>
```

Comment:
I am not convinced that this is a good idea.
Basically it makes the package not "includable" anymore without a specific implementation header on the path.
These kind of "inverse" dependencies should be avoided.

I think we should turn this around: user land code should import the implementation file.
And `rosidl_generator_cpp` should only contain the forward definition as well as a dummy implementation file containing an `#error` statement that lets the user know that he needs a package providing an implementation of this.